### PR TITLE
feat: call trace tree

### PR DESCRIPTION
**Pull request by DanielSchiavini**
- implements #81

### What I did
Allow users to generate a call tree of the last message call

### How I did it
- Add `call_trace` method to the base EVM contract
- Allow trace to be printed or exported to HTML
- Fixed a bug in the ABI parser
- Test the interaction between ABI and Vyper contracts

### How to verify it
- Tests are included

### Description for the changelog
- `Allow users to generate a call tree of the last message call`

### Cute Animal Picture
image

### CODE REVIEW

**Comment by charles-cooper** [comment]
can we split the etherscan caching into its own PR?

**Comment by DanielSchiavini** [comment]
@charles-cooper the trace is pretty unusable without the caching because it's so slow. I can split it though.

File: boa/integrations/jupyter/browser.py [hunk 1]
```diff
@@ -2,11 +2,12 @@
 This module implements the BrowserSigner class, which is used to sign transactions
 in IPython/JupyterLab/Google Colab.
 """
-import json
+
 import logging
 import os
 from asyncio import get_event_loop, sleep
 from itertools import chain
+from json import JSONEncoder, dumps
```
Comment (charles-cooper) [inline boa/integrations/jupyter/browser.py:10]: why?

Comment (DanielSchiavini) [reply boa/integrations/jupyter/browser.py:10]: The json keyword is imported from `boa.rpc` to use the faster json when available. But that does not support custom encoders.

Comment (charles-cooper) [reply boa/integrations/jupyter/browser.py:10]: hmm let's just stick to vanilla `json` module then. i don't think this is particularly performance sensitive

File: boa/contracts/abi/abi_contract.py [hunk 2]
```diff
@@ -361,10 +370,34 @@ def at(self, address: Address | str) -> ABIContract:
         contract = ABIContract(
             self._name, self._abi, self.functions, address, self.filename
         )
-        contract.env.register_contract(address, contract)
+        contract.env.register_contract(contract.address, contract)
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:373]: this change seems unnecessary

File: boa/contracts/trace.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        return f"({', '.join(args)})"
+
+    def _format_output(self, output: bytes):
+        if output == b"":
+            return " => None"
+        decoded = abi_decode(self._output_schema, output)
+        return f" => ({', '.join(_to_str(d) for d in decoded)})"
+
+    @property
+    def _input_schema(self) -> str:  # must be implemented by subclasses
```
Comment (charles-cooper) [inline boa/contracts/trace.py:27]: maybe rename this to `_args_schema`

or at least a comment explaining what it is, took me a couple minutes to figure out what it was

File: boa/contracts/trace.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def _argument_names(self) -> list[str]:  # must be implemented by subclasses
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def _output_schema(self) -> str:  # must be implemented by subclasses
```
Comment (charles-cooper) [inline boa/contracts/trace.py:35]: ditto -- maybe something like `_return_schema`

File: boa/contracts/trace.py (hunk 3 above)
Comment (charles-cooper) [reply boa/contracts/trace.py:27]: in vyper_contract.py, this is called `args_abi_type`

File: boa/contracts/trace.py (hunk 4 above)
Comment (DanielSchiavini) [reply boa/contracts/trace.py]: They don't need to be private I guess, it's pretty public info I think

File: boa/contracts/abi/abi_contract.py [hunk 5]
```diff
@@ -386,15 +386,15 @@ def __repr__(self):
...
+    def args_abi_type(self):
         return f"({_format_abi_type(self.function.argument_types)})"
 
     @cached_property
     def _argument_names(self) -> list[str]:
         return [arg["name"] for arg in self.function._abi["inputs"]]
 
     @cached_property
-    def _output_schema(self):
+    def return_schema(self):
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:397]: i guess for consistency, `return_abi_type`

File: boa/vm/py_evm.py [hunk 6]
```diff
@@ -332,6 +335,30 @@ def finalize(c):
         # swallows exceptions (including Revert).
         return finalize(computation)
 
+    @cached_property
+    def trace(self) -> TraceFrame:
```
Comment (charles-cooper) [inline boa/vm/py_evm.py:339]: maybe rename to `def call_trace(self)`

File: boa/vm/py_evm.py [hunk 7]
```diff
@@ -332,6 +335,30 @@ def finalize(c):
...
+        return self._trace()
+
+    def _trace(self, depth=0) -> TraceFrame:
```
Comment (charles-cooper) [inline boa/vm/py_evm.py:342]: ditto, `_compute_call_trace`

**Review by charles-cooper — COMMENTED** [review]
looks good, just two naming nits

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**